File: iOS/SLC-Lovers/Services/SecureStorage.swift
import Foundation
import Security

/// 安全存储：Keychain 封装
/// 用于存 GitHub Token、加密密钥等敏感信息
enum SecureStorage {

    private static let service = "com.slclovers.app"

    /// 保存
    @discardableResult
    static func save(_ value: String, forKey key: String) -> Bool {
        guard let data = value.data(using: .utf8) else { return false }
        return save(data, forKey: key)
    }

    @discardableResult
    static func save(_ data: Data, forKey key: String) -> Bool {
        // 先删除旧的
        delete(forKey: key)

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        let status = SecItemAdd(query as CFDictionary, nil)
        return status == errSecSuccess
    }

    /// 读取字符串
    static func loadString(forKey key: String) -> String? {
        guard let data = loadData(forKey: key) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// 读取 Data
    static func loadData(forKey key: String) -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess else { return nil }
        return result as? Data
    }

    /// 删除
    @discardableResult
    static func delete(forKey key: String) -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }

    // 业务键
    enum Key {
        static let githubToken = "github.token"
        static let githubOwner = "github.owner"
        static let githubRepo = "github.repo"
        static let pairingCode = "pairing.code"
        static let pairingId = "pairing.id"
        static let lastSyncAt = "sync.lastAt"
    }
}

File: Android/app/src/main/java/com/slclovers/app/sync/SecureStorage.kt
package com.slclovers.app.sync

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/**
 * 安全存储：用 EncryptedSharedPreferences 存敏感信息
 */
object SecureStorage {

    private const val PREFS_NAME = "slc_secure_prefs"

    private fun masterKey(context: Context): MasterKey = MasterKey.Builder(context)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()

    private fun prefs(context: Context) = EncryptedSharedPreferences.create(
        context,
        PREFS_NAME,
        masterKey(context),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    fun save(context: Context, key: String, value: String) {
        prefs(context).edit().putString(key, value).apply()
    }

    fun load(context: Context, key: String): String? = prefs(context).getString(key, null)

    fun delete(context: Context, key: String) {
        prefs(context).edit().remove(key).apply()
    }

    fun clearAll(context: Context) {
        prefs(context).edit().clear().apply()
    }

    // 业务键
    object Key {
        const val GITHUB_TOKEN = "github.token"
        const val GITHUB_OWNER = "github.owner"
        const val GITHUB_REPO = "github.repo"
        const val PAIRING_CODE = "pairing.code"
        const val PAIRING_ID = "pairing.id"
        const val LAST_SYNC_AT = "sync.lastAt"
    }
}